src/elv.py

```diff
@@ -0,0 +1,142 @@
+
+# class for nodes
+class Node():
+    def __init__(self, indexlist, parent, children):
+        self.indexlist = indexlist
+        self.parent = parent
+        self.children = children
+
+    def reproduction(self, child_node, first_letter):
+        self.children[first_letter] = child_node
+        child_node.parent = self
+
+    def index(self):
+        return self.indexlist
+    
+    def update_index(self, index_pos, new_index):
+        self.indexlist[index_pos] = new_index
+    
+    def node_parent(self):
+        return self.parent
+        
+    # if node is leaf int represents start position of suffix in x
+    def is_leaf(self):
+        return type(self.children) == int
+
+    def exists_child(self, letter):
+        return letter in self.children
+    
+    def get_child(self, letter):
+        return self.children[letter]
+    
+    def __str__(self):
+        if self.is_leaf():
+            return f'I am leaf {self.children} with edge {self.indexlist}'
+        else:
+            return f'I am inner node between {self.indexlist}'
+
+
+def construct_tree(x):
+    '''Function construct_tree(x) that takes a string x and constructs a suffix tree 
+    using the class Node.'''
+
+    root = Node(None, None, {})
+    T = [root]
+
+    x += '$'
+
+    for i in range(len(x)): # loop through all suffixes
+        node, head_l, match_l = search_tree(x, T, i) # ab$, T, b$
+        print(node, head_l, match_l)
+        if match_l == 0: # no match in node, extend from it
+            extend_from_node(node, x[i:], x, head_l, T)
```

Comment:
If `x` is a string, what I suspect it is based on line 46, then this is a slice that copies the entirety of `x[i:]`. If you just want to insert a new edge, you want a constant time operation, but this takes time proportional to `len(x[i:])`.